osbuild/util/fscache.py

```diff
@@ -0,0 +1,918 @@
+                if os.path.lexists(path_uuid):
+                    with ctx.suppress_oserror(errno.ENOENT, errno.ENOTDIR):
+                        os.unlink(path_info)
+                    self._update_cache_size(-info["size"])
+        finally:
+            if lockfd is not None:
+                if uuidname is not None:
+                    # In case this runs after the object was renamed, but before
+                    # `uuidname` was cleared, then `_rm_r_object()` will be a
+                    # no-op.
+                    self._rm_r_object(os.path.join(self._dirname_objects, uuidname))
+                linux.fcntl_flock(lockfd, linux.fcntl.F_UNLCK)
+                os.close(lockfd)
+
+    @contextlib.contextmanager
+    def load(self, name: str):
+        """Load a cache entry
+
+        Find the cache entry with the given name, acquire a read-lock and
+        yield its path back to the caller. Once control returns, the entry
+        is released.
+
+        The returned path is the relative path between the cache and the top
+        level directory of the cache entry.
+
+        Parameters:
+        -----------
+        name
+            Name of the cache entry to find.
+        """
+
+        assert self._is_active()
+
+        with contextlib.ExitStack() as es:
+            # Use an ExitStack so we can catch exceptions raised by the
+            # `__enter__()` call on the context-manager. We want to catch
+            # `OSError` exceptions and convert them to cache-misses.
+            try:
+                es.enter_context(
+                    self._atomic_open(
+                        os.path.join(
+                            self._dirname_objects,
+                            name,
+                            self._filename_object_lock,
+                        ),
+                        write=False,
+                        wait=False,
+                    )
+                )
+            except OSError as e:
+                if e.errno in [errno.EAGAIN, errno.ENOENT, errno.ENOTDIR]:
+                    raise self.MissError()
+                raise e
+
+            yield os.path.join(
+                self._dirname_objects,
+                name,
+                self._dirname_data,
+            )
+
+    def extract(self, name: str, path: str):
+        """Copy cache entry into local directory
+
+        Look up the cache entry with the given name and copy its entire `data`
+        directory into `path`. If `name` is not found, a `MissError`
+        exception is raised.
+
+        Parameters:
+        -----------
+        name
+            Name of the cache entry to find.
+        path
+            Target path where to copy the cache entry to.
+        """
+
+        with self.load(name) as path_data:
+            self._copy_a(path_data, path)
+
+    def configure(
+        self,
+        creation_boot_id: Optional[str] = None,
+        maximum_size: Optional[int] = None,
```

Comment:
Can we have a getter for the size too, please? ;)

Reply:
I am tempted to turn this into a property that takes a named-tuple as argument. So you can query the `info` field to get the named-tuple of the current configuration, or set it to effectively get what this function-call does. Does that sounds ok? Or would you prefer individual getters and setters?

Reply:
Sounds good!